pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,346 @@
+		})
+	}
+}
+
+func Test_New_Happy(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name      string
+		client    *http.Client
+		installID string
+		want      Client
+	}{
+		{
+			name: "happy_path_no_install_id",
+			want: defaultClient(),
+		},
+		{
+			name:      "happy_path_with_install_id",
+			installID: testInstallID,
+			want:      defaultClient(),
+		},
+		{
+			name:      "happy_path_with_custom_http_client",
+			installID: testInstallID,
+			client:    &http.Client{Timeout: 2},
+			want: func() Client {
+				c := defaultClient()
+				c.httpClient = &http.Client{Timeout: 2}
+				return c
+			}(),
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+
+			installPath := t.TempDir() + "/" + installIDFileName
+			if len(tc.installID) > 0 {
+				if err := storeInstallID(testAppID, installPath, &InstallIDData{tc.installID}); err != nil {
+					t.Fatalf("test setup failed: %s", err.Error())
+				}
+			}
+			envVars := map[string]string{
+				"METRICS_URL": testServerURL,
+			}
+			lookupper := envconfig.MapLookuper(envVars)
+			opts := make([]Option, 0, 2)
+			opts = append(opts, WithLookuper(lookupper))
+			opts = append(opts, withInstallIDFileOverride(installPath))
+			if tc.client != nil {
+				opts = append(opts, WithHTTPClient(tc.client))
+			}
+
+			got, err := New(ctx, testAppID, testVersion, opts...)
+
+			storedID, err := loadInstallID(testAppID, installPath)
+			if err != nil {
+				t.Fatalf("could not load install ID for checking side effects")
+			}
+			if len(tc.installID) > 0 {
+				if diff := cmp.Diff(storedID.InstallID, tc.installID); diff != "" {
+					t.Errorf("install id changed. Diff (-got +want): %s", diff)
+				}
+			} else if storedID.InstallID == "" {
+				t.Errorf("install id not saved")
+			} else {
+				// We cannot know ahead of time if generated, so copy from got to want.
+				tc.want.installID = got.installID
+			}
+			if diff := cmp.Diff(got.installID, storedID.InstallID); diff != "" {
+				t.Errorf("install id in client does not match stored. Diff (-client +stored): %s", diff)
+			}
+
+			if diff := cmp.Diff(got.appID, tc.want.appID); diff != "" {
+				t.Errorf("unexpected app id. Diff (-got +want): %s", diff)
+			}
+			if diff := cmp.Diff(got.version, tc.want.version); diff != "" {
+				t.Errorf("unexpected version. Diff (-got +want): %s", diff)
+			}
+			if diff := cmp.Diff(got.installID, tc.want.installID); diff != "" {
+				t.Errorf("unexpected install id. Diff (-got +want): %s", diff)
+			}
+			if diff := cmp.Diff(got.httpClient, tc.want.httpClient); diff != "" {
+				t.Errorf("unexpected httpClient. Diff (-got +want): %s", diff)
+			}
+			if diff := cmp.Diff(got.optOut, tc.want.optOut); diff != "" {
+				t.Errorf("unexpected optOut. Diff (-got +want): %s", diff)
+			}
+			if diff := cmp.Diff(got.config, tc.want.config); diff != "" {
+				t.Errorf("unexpected config. Diff (-got +want): %s", diff)
+			}
+		})
```

Comment:
Couldn't this all be replaced with `cmp.Diff` on `got` and `want`? If not, we should restructure the test to enable that.

Reply:
The issue was with unexported fields (`cmp.Diff` can't investigate them).

I could make public, was just hoping to reduce the leakiness of my abstraction.

Reply:
You can have public fields in private structures, which prevents the abstraction from leaking

Reply:
https://github.com/abcxyz/abc-updater/pull/61